homeassistant/components/husqvarna_automower/calendar.py

```diff
@@ -0,0 +1,114 @@
+"""Creates a calendar entity for the mower."""
+
+from datetime import datetime
+import logging
+
+from aioautomower.model import AutomowerCalendarEvent
+from ical.calendar import Calendar
+from ical.event import Event
+from ical.types.recur import Recur
+
+from homeassistant.components.calendar import CalendarEntity, CalendarEvent
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import AutomowerConfigEntry
+from .coordinator import AutomowerDataUpdateCoordinator
+from .entity import AutomowerBaseEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: AutomowerConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up lawn mower platform."""
+    coordinator = entry.runtime_data
+    async_add_entities(
+        AutomowerCalendarEntity(mower_id, coordinator) for mower_id in coordinator.data
+    )
+
+
+class AutomowerCalendarEntity(AutomowerBaseEntity, CalendarEntity):
+    """Representation of the Automower Calendar element."""
+
+    _attr_name: str | None = None
+
+    def __init__(
+        self,
+        mower_id: str,
+        coordinator: AutomowerDataUpdateCoordinator,
+    ) -> None:
+        """Set up AutomowerCalendarEntity."""
+        super().__init__(mower_id, coordinator)
+        self._attr_unique_id = mower_id
+        self.calendar = Calendar()
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the current or next upcoming event."""
+        self._automower_to_ical_event(self.mower_attributes.calendar.events)
+        return _ical_to_calendar_event(self.calendar.events[0])
```

Comment:
Is there a promise here that the first event in the list is the next upcoming event? I'm worried about a case like event[0] is {monday, friday} and event[1] is {wedensday} and its currently tuesday.

Even if its returned from the API in some specific order, there may be a lag between when events are fetched and the current clock changing the state of which event is next.  I would expect you need to do something similar to google calendar here: https://github.com/home-assistant/core/blob/769d7214a31bbb6df9ae9d09832e843e18016dfc/homeassistant/components/local_calendar/calendar.py#L96

where you need to re-evaluate the entire timeline and decide which event is actually next.

Reply:
Maybe https://github.com/home-assistant/core/blob/769d7214a31bbb6df9ae9d09832e843e18016dfc/homeassistant/components/rainbird/calendar.py#L71 is a better example since its a similar in that it fetches a schedule from a device.